### PR TITLE
Add Recurring support and update latest lib

### PR DESCRIPTION
**Pull request by dzungtran**
- Add Recurring support
- Update latest lib

### CODE REVIEW

File: src/modules/gateways/callback/paymentwall.php [hunk 1]
```diff
@@ -27,79 +28,108 @@
...
-
-        }
+        processDeliverable($invoiceid, $pingback, $gateway, $userData, $orderData);
     } elseif ($pingback->isCancelable()) {
+        // WHMCS not supported
         logTransaction($gateway["name"], $_GET, "Not Supported");
+        die("Not Supported");
+    } else {
+        switch ($pingback->getType()) {
+            case Paymentwall_Pingback::PINGBACK_TYPE_SUBSCRIPTION_EXPIRED:
```
Comment (sevastyanovio) [inline src/modules/gateways/callback/paymentwall.php:40]: break?

File: src/modules/gateways/callback/paymentwall.php [hunk 2]
```diff
@@ -27,79 +28,108 @@
...
+    $invoiceItems = select_query(
+        'tblinvoiceitems',
+        '*',
+        array("invoiceid" => $invoiceid));
+
+    $hostId = false;
+    $hosting = false;
+    // Get invoice products
+    while ($item = mysql_fetch_assoc($invoiceItems)) {
+        if ($item['relid'] != 0 && $item['type'] == 'Hosting') {
```
Comment (sevastyanovio) [inline src/modules/gateways/callback/paymentwall.php]: let's store types in consts
or trying to get from CMS

File: src/modules/gateways/callback/paymentwall.php [hunk 3]
```diff
@@ -27,79 +28,108 @@
...
+            $hostId = $item['relid'];
+        }
+    }
+
+    if ($hostId) {
+        $hosting = mysql_fetch_assoc(select_query(
+            'tblhosting',
+            'tblhosting.id,tblhosting.username,tblproducts.autosetup,tblhosting.packageid',
+            array("tblhosting.id" => $hostId), false, false, 1,
```
Comment (sevastyanovio) [inline src/modules/gateways/callback/paymentwall.php]: What is `false, false, 1`?

Better to have some description/consts for that fields.

File: src/modules/gateways/callback/paymentwall.php [hunk 4]
```diff
@@ -27,79 +28,108 @@
...
+            "tblproducts ON tblhosting.packageid=tblproducts.id"
+        ));
+    }
+
+    // Update subscription id
+    if ($hosting) {
+        update_query('tblhosting', array('subscriptionid' => $pingback->getReferenceId()), array('id' => $hosting['id']));
+    }
+
```
Comment (sevastyanovio) [inline src/modules/gateways/callback/paymentwall.php:91]: move to `updateSubscriptionId()` method

File: src/modules/gateways/callback/paymentwall.php [hunk 5]
```diff
@@ -27,79 +28,108 @@
...
+    // Check enable delivery request
+    if (isset($gateway['enableDeliveryApi']) && $gateway['enableDeliveryApi'] != '' && $hosting) {
```
Comment (sevastyanovio) [inline src/modules/gateways/callback/paymentwall.php]: '' is false
so, instead `$gateway['enableDeliveryApi'] != ''`
just `$gateway['enableDeliveryApi']`

File: src/modules/gateways/paymentwall.php [hunk 6]
```diff
@@ -25,37 +28,111 @@ function init_paymentwall_config($params)
...
+        $onetimeWidget = new Paymentwall_Widget(
+            $params['clientdetails']['email'],
+            $params['widget'],
+            array(
+                $onetimeProduct
+            ),
+            array_merge(
+                array(
+                    'integration_module' => 'whmcs',
+                    'is_test' => (isset($params['isTest']) && $params['isTest'] != '') ? 1 : 0,
```
Comment (sevastyanovio) [inline src/modules/gateways/paymentwall.php]: just `isset($params['isTest'])`

`true` will be converted to `1` and `false` to `0`

File: src/modules/gateways/paymentwall.php [hunk 7]
```diff
@@ -25,37 +28,111 @@ function init_paymentwall_config($params)
...
+                'width' => '100%',
+                'height' => '500'
+            ))) . '" />
+            <input type="image" src="' . $params['systemurl'] . '/images/paymentwall/subscribe.png" onClick="this.form.submit()"/>
     </form>';
+    }
 
+    if ((!$params['forceSubscription'] && $recurring) || !$recurring) {
+        $onetimeWidget = new Paymentwall_Widget(
+            $params['clientdetails']['email'],
```
Comment (sevastyanovio) [inline src/modules/gateways/paymentwall.php]: can we send user id? instead email

also better to send email in `email` parameter, not in `uid`

https://www.paymentwall.com/en/documentation/Digital-Goods-API/710#paymentwall_widget_call_flexible_widget_call

File: src/modules/gateways/paymentwall.php [hunk 8]
```diff
@@ -25,37 +28,111 @@ function init_paymentwall_config($params)
...
+                $recurring['firstpaymentamount'],
                 $params['currency'],
-                $params["description"],
-                Paymentwall_Product::TYPE_FIXED
-            )
-        ),
-        array_merge(
+                $params["description"] . ' + Setup Fee',
+                Paymentwall_Product::TYPE_SUBSCRIPTION,
+                1,
```
Comment (sevastyanovio) [inline src/modules/gateways/paymentwall.php]: what if subscription for 3 mounts?

File: src/modules/gateways/paymentwall.php [hunk 9]
```diff
@@ -25,37 +28,111 @@ function init_paymentwall_config($params)
...
+    $code = '';
+    $hasTrial = false;
+    $subscriptionProduct = false;
+    $subscriptionWidget = false;
+    $onetimeProduct = false;
+    $onetimeWidget = false;
+
+    if ($recurring) {
+        $trialProduct = null;
+        $periodType = ($cycleUnits == 'Y') ? Paymentwall_Product::PERIOD_TYPE_YEAR : Paymentwall_Product::PERIOD_TYPE_MONTH;
```
Comment (sevastyanovio) [inline src/modules/gateways/paymentwall.php]: move to `getPeriodType` method

File: src/modules/gateways/paymentwall.php [hunk 10]
```diff
@@ -25,37 +28,111 @@ function init_paymentwall_config($params)
...
+    }
+
+    $onetimeProduct = new Paymentwall_Product(
+        (int)$params['invoiceid'],
+        $params['amount'],
+        $params['currency'],
+        $params["description"],
+        Paymentwall_Product::TYPE_FIXED
+    );
+
```
Comment (sevastyanovio) [inline src/modules/gateways/paymentwall.php]: move to `getOneTimeProduct` method

File: src/modules/gateways/paymentwall.php [hunk 11]
```diff
@@ -25,37 +28,111 @@ function init_paymentwall_config($params)
...
+            $recurring['recurringamount'],
+            $params['currency'],
+            $params["description"],
+            Paymentwall_Product::TYPE_SUBSCRIPTION,
+            $recurring['recurringcycleperiod'],
+            $periodType,
+            true,
+            $trialProduct
+        );
+    }
```
Comment (sevastyanovio) [inline src/modules/gateways/paymentwall.php:45]: move to `getSubscriptionProduct` method

File: src/modules/gateways/callback/paymentwall.php (hunk 4 above)
Comment (dzung2t) [inline src/modules/gateways/callback/paymentwall.php]: 2 functions update sub id not same the conditions.
```php 
update_query('tblhosting', array('subscriptionid' => $pingback->getReferenceId()), array('id' => $hosting['id']));
```

```php 
update_query('tblhosting', array('subscriptionid' => ''), array('subscriptionid' => $pingback->getReferenceId()));
```

**Comment by dzungtran** [comment]
I've made some changes in source codes. 
Please review it again.

File: src/modules/gateways/callback/paymentwall.php [hunk 12]
```diff
@@ -27,79 +29,120 @@
...
+                'status' => 'sent',
+                'reference_id' => $pingback->getReferenceId(),
+                'invoice_id' => $invoiceid,
+                'data' => json_encode($data),
+                'updated_date' => time(),
+            ), array(
+                'id' => $deliveryData['id']
+            ));
+        }
+
```
Comment (sevastyanovio) [inline src/modules/gateways/callback/paymentwall.php:138]: move to sendDeliveryApiRequest() method

File: src/modules/gateways/callback/paymentwall.php [hunk 13]
```diff
@@ -27,79 +29,120 @@
...
+        $hosting = mysql_fetch_assoc(select_query(
+            'tblhosting', // table name
+            'tblhosting.id,tblhosting.username,tblproducts.autosetup,tblhosting.packageid', // fields name
+            array("tblhosting.id" => $hostId), // where conditions
+            false, // order by
+            false, // order by order
+            1, // limit
+            "tblproducts ON tblhosting.packageid=tblproducts.id" // join
+        ));
+    }
```
Comment (sevastyanovio) [inline src/modules/gateways/callback/paymentwall.php:85]: getHostId() method

**Pull request merged**